File: 백준/Gold/12865. 평범한 배낭/평범한 배낭.py
from sys import stdin   

n, k = map(int, stdin.readline().split())

values = [list(map(int, stdin.readline().split())) for _ in range(n)]

dp = [[0]*(k+1) for _ in range(n+1)]
for i in range(1, n+1):
    w,v  = values[i-1]

    for j in range(1, k+1):
        if j < w:
            dp[i][j] = dp[i-1][j]
        else:
            dp[i][j] = max(dp[i-1][j-w]+v, dp[i-1][j])
print(dp[n][k])
    

File: 백준/Gold/12865. 평범한 배낭/평범한 배낭.java

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {

    public static void main(String[] args) throws NumberFormatException, IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int n = Integer.parseInt(st.nextToken());
        int k = Integer.parseInt(st.nextToken());

        int[] dp = new int[k+1];

        for (int i = 0; i < n; i++)
        {
            st = new StringTokenizer(br.readLine());
            int weight = Integer.parseInt(st.nextToken());
            int value = Integer.parseInt(st.nextToken());

            for (int idx = k; idx >= weight; idx--){
                dp[idx] = Math.max(dp[idx-weight]+value, dp[idx]);
            }
        }

        System.out.println(dp[k]);

    }
}
